lfs/gitscanner_log.go

```diff
@@ -63,6 +64,106 @@ func scanUnpushed(cb GitScannerFoundPointer, remote string) error {
 	return nil
 }
 
+func scanStashed(cb GitScannerFoundPointer, s *GitScanner) error {
+
+	// Stashes are actually 2-3 commits, each containing one of:
+	// 1. Working copy modified files
+	// 2. Index changes
+	// 3. Untracked files, if -u was used
+	// We need to get the diff of all 3 of these commits to ensure we have all
+	// of the LFS objects necessary to pop the stash
+
+	// First get the list of stashes
+	// In recent version of git you can get parents directly from
+	// this command, avoiding the intermediate "git show"
+	// However older gits (at least <=2.7) don't report merge parents in the reflog
+	// So we need to do it in 2 stages
+	logArgs := []string{"-g", "--format=%h", "refs/stash", "--"}
+
+	cmd, err := git.Log(logArgs...)
+	if err != nil {
+		return err
+	}
+
+	scanner := bufio.NewScanner(cmd.Stdout)
+
+	var allStashShas []string
+	for scanner.Scan() {
+		leafSha := strings.TrimSpace(scanner.Text())
+
+		allStashShas = append(allStashShas, leafSha)
+
+		// For each leaf, use "git show" to expand parents & thus get
+		// all 2-3 shas involved in the stash
+		// As mentioned newer gits could do this in the reflog output but not gteed
+		showArgs := []string{"--quiet", "--format=%p", leafSha}
+		showCmd, err := git.Show(showArgs...)
+		if err != nil {
+			return err
+		}
+
+		// gets the abbreviated parent hashes as :
+		//      A B [C]
+		// A = Parent commit of the stash (ignore, not part of the stash)
+		// B = Index changes for the hash
+		// C = Untracked files (optional, only present if -u)
+		showScanner := bufio.NewScanner(showCmd.Stdout)
+
+		for showScanner.Scan() {
+			line := strings.TrimSpace(showScanner.Text())
+			refs := strings.Split(line, " ")
+			for i, ref := range refs {
+				if i > 0 { // Extra merge parents
+					allStashShas = append(allStashShas, ref)
+				}
+			}
+		}
+		err = showCmd.Wait()
+		if err != nil {
+			return err
+		}
+
+	}
+	err = cmd.Wait()
+	if err != nil {
+		// Ignore this error, it really only happens when there's no refs/stash
+		return nil
+	}
+
+	// Now we need to specifically use "git show" to parse results
+	// We can't use "git log" because weirdly that omits the index changes
+	// in the diff display, it collapses both into one diff and only shows the
+	// final change (not a 3-way like show). Only "show" on all the shas
+	// above displays them separately
```

Comment:
Please forgive me if I've got the wrong end of the stick here!  If I'm following along, we use `git log -g` to get the reflog list of stashes; for each stash, this is the final `WIP` merge commit, which may have additional parents in the form of the index and untracked-files commits.

So we use `git show` to retrieve the SHAs of each of those, for each stash, and then iterate again over that last running `git show` on the whole concatenated list of SHAs, in order to get any LFS pointer data out of them.

I can't help but feel that there might be a more compact way to achieve the same results.  I haven't tried this in code yet, but what about using `git log -m` plus the usual `logLfsSearchArgs` (i.e., `-p` in particular) just each of the SHAs from the first list from the reflog, of the WIP merge commits?  Doing that seemed like it might get all the necessary LFS data from all of the parents in one go, without needing run `git show` at all.

I realize older Git versions may throw a wrinkle in here, but testing with Git v2.0.0 it seems like it might work (again, on the command line only, not in code yet).

I can try to write up another version if you don't get to it first, just to test this out ... but I may not get it done tomorrow, so I won't promise that!  :-)

Reply:
If you can find a better way to do it I'm quite happy! I kind of ran out of time with this, I spent way longer than I expected last week tinkering with it and this way was the only one I came up with in the time that worked consistently. I did try a few log variations but there were 2 problems:

1. `log -g` doesn't show merges in older git versions
1. In my experiments `log` often collapsed the diff of the WIP and index, making it impossible to get all 3 changes. I don't know why it did this, but only `show` reliably gave me all 3 OIDs. If you can find a way to make it work all the time, great!

Reply:
Ah sounds like from your comments the `-m` option might fix this and the more awkward diff format. That will be much better!

I don't know when I'm going to have time to revisit this though, it'll probably be a few days.

Reply:
No worries at all; you've gotten this a huge distance down the road and much faster than I could have.  Let me see if I can find some time in the next day or two to tinker a bit more with this.

Would it be OK if I pushed commits onto your branch in your repo?  I haven't wanted to do that without asking for permission first (even if the "allow edits from maintainers" checkbox is checked on your side).

Reply:
Oh yeah that's fine, thanks for asking but I'm very happy to collaborate on this. :)

Reply:
I circled back around to this (thanks for your patience!) and put together 7fb982f946cdd0db9ec0e3a2d96285be41a2b129, which seems to pass the new test suite and lets us use `git log` exclusively again, I think.

Reply:
Thanks for doing that, it's a much better solution. We win! 👍 😄